GUIDELINES.md

```diff
@@ -0,0 +1,135 @@
+  * Always use the same package structure for test sources, e.g. if you app resides in a module `app` and has a base package `org.myorg.product.app` then put the module test configuration in exactly in the same package
+  * If there is a class `org.myorg.product.app.view.LoginViewModel` then the test class `LoginViewModelTest` and `LoginViewModelSteps` would reside in the exact same package, too
+  * If you have a dedicated module for test sources you should still stick to the same package structure, e.g. if the module is `:app:test` _don't_ add `test` to the package structure for test sources
+  * If modules depend on each other the module testing configuration should do the same, e.g. `val appModuleTestingConfiguration = moduleTestingConfiguration(dependentConfig1, dependentConfig2, ...) { ... }`, this automatically imports all dependent configurations
+* Create a test class for the _feature under test_ (e.g. `LoginTest`), so the test is agnostic of the underlying structures (classes, models, views, data structures, ...)
+  * Should it be that a test really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelTest`)
+* Create a steps class for the feature under test (e.g. `LoginSteps`) which will contain all the `given`, `when` and `then` functions as well as dependency configuration and setup code
+  * The steps class should know about the setup of the test, so for example if the steps class tailored to test the integration of multiple classes, it should also be called something like `LoginIntegrationSteps` which tests a broad stack of classes (e.g. a `LoginViewModel` talking to an `AuthManager` and a `SessionStore`)
+  * Should it be that a steps class really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelSteps`)
+  * In case the steps class' aim is purely to offer and act on a fake implementation of a class, name it with a `FakeSteps` suffix (e.g. `BackendGatewayFakeSteps` or `SessionStoreFakeSteps`)
+* Start with the test
+  * In the test class you created add reference to the steps class with the `steps` function (e.g. `val sut by steps<LoginSteps>`) and call the value `sut` (system under test)
+  * Fill up the test with `@Test`-annotated test functions
+    * The test function can be named after a _scenario_ (e.g. `New user trying to log in`, `Weather condition is cloudy`)
+    * Only if necessary to distinguish test cases, also add more context like what's roughly expected (e.g. `New user trying to login - welcome to the app message shown`, `Existing user trying to login - welcome back message shown`)
+    * Don't use names like `givenNewUser_whenLoggingIn_thenShowWelcomeToAppMessage` because
```

Comment:
I agree, that it shouldn't be written in one long word. But written as sentence, this can imo be a valid test name

Reply:
i think it's a redundancy to repeat yourself in the name of the test function _and_ in the test body. when the test is properly abstracted (preferrably around business domain concepts) you'll find exactly these functions in the body:

```
givenNewUser()
whenLoggingIn()
thenShowWelcomeToAppMessage()
```

as the test body is so easy to read i think it becomes obsolete to be that precise in the function name. see also arguments in the comment above...

Reply:
maybe a hint that the abstracted style inside the test functions helps creating more context for the test to analyze is enough here then instead of discouraging the use of long function names.